blockstore/apps/api/v1/serializers/drafts.py

```diff
@@ -0,0 +1,109 @@
+"""
+Serializers for Draft (the Django Model) and StagedDraft (the repo primitive)
+"""
+import base64
+import binascii
+
+from django.core.files.base import ContentFile
+from rest_framework import relations, serializers
+from rest_framework.reverse import reverse
+
+from blockstore.apps.bundles.models import Bundle, Draft
+from blockstore.apps.bundles.store import DraftRepo, SnapshotRepo
+
+
+class StagedDraftField(serializers.Field):
+    """Read-only field for a StagedDraft."""
+
+    def to_representation(self, staged_draft):
+        """StagedDraft JSON serialization."""
+        draft_repo = DraftRepo(SnapshotRepo())
+        if staged_draft.base_snapshot is None:
+            base_snapshot_repr = None
+        else:
+            base_snapshot_repr = staged_draft.base_snapshot.hash_digest.hex()
+
+        basic_info = {
+            'uuid': staged_draft.uuid,
+            'bundle_uuid': staged_draft.bundle_uuid,
+            'name': staged_draft.name,
+            'base_snapshot': base_snapshot_repr,
+            'created_at': staged_draft.created_at,
+            'updated_at': staged_draft.updated_at,
+        }
+
+        basic_info['files'] = {
+            path: {
+                "url": draft_repo.url(staged_draft, path),
+                "size": file_info.size,
+                "hash_digest": file_info.hash_digest.hex(),
+                "modified": path in staged_draft.files_to_overwrite
+            }
+            for path, file_info in staged_draft.files.items()
+        }
+
+        return basic_info
+
+
+class DraftSerializer(serializers.ModelSerializer):
+    """Serializer for Drafts, appropriate for list views. No files metadata."""
+    class Meta:
+        model = Draft
+        queryset = Draft.objects.all().select_related('bundle')
+        fields = ('uuid', 'url', 'bundle', 'name')
+
+    url = relations.HyperlinkedIdentityField(
+        lookup_field='uuid',
+        view_name='api:v1:draft-detail',
+    )
+    bundle = relations.SlugRelatedField(
+        slug_field='uuid',
+        queryset=Bundle.objects.all(),
+    )
+
+class DraftWithFileDataSerializer(DraftSerializer):
+    """
+    DraftSerializer subclass that adds files metadata.
+
+    This payload could become very large, so this serializer should not be used
+    for list views.
+    """
+    class Meta:
+        model = Draft
+        queryset = Draft.objects.all()
+        fields = DraftSerializer.Meta.fields + ('staged_draft',)
+
+    staged_draft = StagedDraftField()
+
+
+class DraftFileUpdateSerializer(serializers.BaseSerializer):
+    """Write-only serializer for Draft file updates."""
+
+    def to_internal_value(self, data):
+        """
+        Convert file names -> b64 strings into file names -> ContentFiles.
+
+        `data` is a  dict of file names to base64 encoded data representations.
```

Comment:
How does one delete a file from a draft? (so it exists in the bundle, but I want it to not exist once the draft is committed)

How does one reset a files changes in the draft? (so the file now has `"modified": false` but doesn't have to be re-uploaded via `PATCH`. Fine if this is added later, but I think deleting files is needed now.

Reply:
Delete is PATCHing with that filename set to null.

No facility for resetting at the moment.